0000-ilp-over-http.md

```diff
@@ -0,0 +1,57 @@
+---
+title: ILP Over HTTP
+draft: 1
+---
+
+# ILP Over HTTP
+> A bilateral communication protocol for server-to-server connections
+
+## Motivation
+
+Scaling Interledger infrastructure to handle large volumes of ILP packets requires horizontally scaling connectors. Using HTTP for bilateral communication enables service providers to leverage standard tools and services for hosting, load balancing, Distributed Denial of Service (DDoS) protection, and monitoring.
+
+## Overview
+
+In an ILP Over HTTP connection, both peers run HTTP servers with accessible HTTPS endpoints. When peering, the peers exchange their respective URLs, authentication tokens or TLS certificates, ILP addresses, and settlement-related details.
+
+Each ILP Prepare packet is sent as the body of an HTTP request to the peer's server endpoint. The HTTP response body will contain either the ILP Fulfill or Reject packet.
+
+## Specification
+
+This is a minimal protocol built on HTTP. HTTP/2 is RECOMMENDED for performance reasons, although HTTP/1.1 MAY also be used. Implementations SHOULD support HTTP version negotiation via Application Protocol Negotiation (ALPN).
+
+### Authentication
+
+Peers MAY use any standard HTTP authentication mechanism to authenticate incoming requests. Bearer Tokens or TLS Client Certificates are RECOMMENDED.
+
+### Request
+
+```http
+POST /ilp HTTP/2.0
+Accept: application/octet-string
+Content-Type: application/octet-string
+Authorization: Bearer zxcljvoizuu09wqqpowipoalksdflksjdgxclvkjl0s909asdf
+Host: connector.example
+X-ILP-Destination: example.usd.connector.account
+
+< Body: Binary OER-Encoded ILP Prepare Packet >
+```
+
+**Field Details:**
+
+- **Path** - A connector MAY specify any HTTP path for their peer to send ILP packets to.
+- **X-ILP-Destination Header** - (Optional) It is RECOMMENDED to set this header to the destination ILP address from the ILP packet. This enables standard HTTP load balancers to route packets to particular connector instances without parsing the ILP packet.
```

Comment:
Can we also add `ILP-Peer-Name` as in https://github.com/interledgerjs/ilp-plugin-http ? It can be a useful feature in a situation where a stateful cluster of connectors is talking to a stateless cluster of connectors.

Reply:
This is (I think) a symptom of the problem with the current plugin/connector architecture.

The connector assumes that a plugin represents a single account BUT a plugin uses protocols like HTTP and WebSockets where it hosts a server and could have many clients.

The result is you have to have mini-accounts (i.e. a lite connector) implemented inside the plugin.

The correct way to do this is to authenticate a connection and then you can assume that all messages on that connection are from the same client (who has now been authenticated as a specific peer).

In lieu of this I think Ben is suggesting that a peer can identify themselves using this extra header.

Reply:
Unless this type of header is strictly necessary, I'd prefer to leave this to the authentication mechanism chosen. If a peer wants to identify themselves, a username or a specific TLS client cert seems like a better way to do that.